feat: add gitlab cicd deep dive post (#416)
hendriknielaender/double-trouble: data/blog/gitlab-job-analysis.md

> ---

Reviewer: SEO: maybe we could rename the file for better seo. `mastering-gitlab-metrics-sqlite-grafana-guide.md`

Author: Definitely not mastering as this was just our own first attempt.

Maybe: analyzing-gitlab-metrics-sqlite-grafana-guide.md

> publishDate: "Mar 10 2024"

Reviewer: ```suggestion
publishDate: "Mar 12 2024"
```

title: "How to dive deep into Gitlab Metrics with SQLite and Grafana"

> description: "Follow along as we showcase a recent dive into gitlab jobs metrics."

Reviewer: SEO:
Unlock the full potential of GitLab jobs metrics using SQLite and Grafana for deeper insights into pipeline performance.

> description: "Follow along as we showcase a recent dive into gitlab jobs metrics."

Reviewer: ```suggestion
description: "Explore the comprehensive guide to unlocking insights into GitLab jobs metrics using SQLite and Grafana."
```

Author: "Uncover the secrets" is too much and doesnt fit to the content. It is not just a list of techniques, it is telling a story of our own approach and our learnings. I think its good to do SEO estimation but we dont need to clickbait for a blogpost that is solid.

Better:
"Follow along our recent analysis of GitLab jobs metrics using SQLite and Grafana with our step-by-step guide. Enhance your CI/CD pipeline analysis today."

image: "~/assets/images/thumbnails/gitlab-job-analysis.jpg"
imageCreditUrl: https://labs.openai.com

> tags: [GitLab, CI/CD, DevOps, pipeline, automation]

Reviewer: ```suggestion
tags: [GitLab, CI/CD, DevOps, pipeline, automation, Grafana]
```

---

A recent optimization has left us again excited - the pipeline speed for a new project had
significantly improved. Execution times had basically cut in half across the board. Looking at a
few jobs we could confirm this on an individual basis, but we were wondering about the bigger
picture. How had the pipeline improved over a bigger timespan?

> ## GitLabs Dashboards

Reviewer: ```suggestion
## GitLab Dashboards
```

Going into this questions the natural start were the existing GitLab dashboards. GitLab has a wide

> variety of different dashboards build-in, so it seemed likely that we would find the answer within

Reviewer: ```suggestion
variety of different dashboards built-in, so it seemed likely that we would find the answer within
```

GitLab itself.

### CI/CD Analytics

> Under the projects `Analyze` tab, there is the `CI/CD Analytics` dashboard. A basic view of the

Reviewer: ```suggestion
Within the `Analyze` tab, we explored the `CI/CD Analytics` dashboard, offering a glimpse into pipeline run success rates. A basic view of the
```

number of successful and total pipeline runs. Below are the pipelines for the most recent
commit. Pipelines which fail early appear as shorter, which makes it not very useful for our
analysis.

![](../../src/assets/images/posts/gitlab-job-analysis/gitlab-cicd-analytics.png)

Other dashboards in the `Analyze` tab sound intriguing, but they mostly display graphs based on

> the amount of commits, merge requests, GitLab issues or lines changed.

Author: Jo @hendriknielaender i capitalized all instances of gitlab, grafana and sqlite since they are proper nouns. Now this looks really weird though, what do you think? Otherwise this PR is ready from my side.

Reviewer: It is what it is, that's their branding. 🤷‍♂️
I would leave it as it is.

### Build

> In the `Build` tab we can browse through a list of recent jobs and pipelines. The results are simply

Reviewer: ```suggestion
The `Build` tab allowed us to review recent jobs and pipelines, though it lacked comprehensive trend analysis capabilities. The results are simply
```

returned as a paginated table. For pipelines this can give an impression on an individual basis,
but grasping the overall trend over a longer period is hard.

Especially with the Jobs search getting a job overview is challenging. This search is infamous for
only allowing to filter by job status, not by name. The related issue
[#22027](https://gitlab.com/gitlab-org/gitlab/-/issues/22027) for the underlying api is 5 years
old, with many people wishing for its resolution.

## Grafana

> Moving on from the GitLab metrics we turned our eyes towards Grafana.

Reviewer: ```suggestion
Seeking deeper insights, we shifted our focus to integrating Grafana with GitLab metrics.
```

Author: Actually we got no almost no insights from the gitlab metrics, so seeking deeper insights is going too far. These are pretty basic insights which gitlab is lacking.

together with a number of HTTP status codes returned by the runners.

![](https://raw.githubusercontent.com/mvisonneau/gitlab-ci-pipelines-exporter/main/docs/images/grafana_dashboard_jobs.jpg)

The related
[gitlab-ci-pipelines-exporter](https://github.com/mvisonneau/gitlab-ci-pipelines-exporter/tree/main)
in the screenshot above operates at the same level of the GitLab runner, but publishes many more
metrics, including the much desired job information. If you have your runners self-hosted on
kubernetes, together with a prometheus backend then this is probably the best option.

## Grafana and SQLite

With the easy options off the table there were few choices left. We had to build something
ourselves, fetching the data via the GitLab api and ingesting them into some datasource.

The company that we work at does not have a Grafana server and the only monitoring tool is
basically cloudwatch, to ingest this data we will have to start Grafana locally.

The complete code is also available
[here](https://gist.github.com/flyck/e3deb7db07a5817dfb3a5c49b205a1c4).

### 1) Exploring the data

Gladly GitLab offers a supergraph, which can be easily explored with their public [graphql

> explorer](https://gitlab.com/-/graphql-explorer). After a tries we had our paginated query:

Reviewer: ```suggestion
explorer](https://gitlab.com/-/graphql-explorer). After a few tries, we had our paginated query:
```

will need to call this function a couple of times until we hit our threshold.
```typescript
import SQLite from 'bun:sqlite';
import file from 'bun';

// Environment variables for configuration
const REPO_PATH = 'path/to/repo';
const GITLAB_TOKEN = Bun.env.GITLAB_TOKEN;
const PAGE_SIZE = 100;
const JOB_LIMIT = 20000;
const DB_PATH = 'jobs_database.sqlite';
const GRAPHQL_ENDPOINT = 'https://gitlab.com/api/graphql';

async function fetchJobs(after?: string): Promise<{ jobs: JobData[]; nextCursor?: string }> {
  const query = `your GraphQL query here, including variables for pagination`;
  const variables = after ? { after, first: PAGE_SIZE } : { first: PAGE_SIZE };

  const response = await fetch(GRAPHQL_ENDPOINT, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${GITLAB_TOKEN}`,
    },
    body: JSON.stringify({
      query,
      variables,
    }),
  });

  const { data } = await response.json() as any;
  const jobs: JobData[] = data.project.jobs.nodes;
  const nextCursor = data.project.jobs.pageInfo.endCursor;

  return { jobs, nextCursor };
}
```

### 3) Transforming the data

> After we downloading the data, we can dump it into a SQLite database for local analysis. But

Reviewer: ```suggestion
After downloading the data, we can dump it into a SQLite database for local analysis. But
```

    pipelineComputeMinutes REAL,
    pipelineDuration INTEGER,
    pipelineComplete INTEGER
  );
`).run();

function insertJobs(jobs: JobData[]) {
  jobs.forEach(job => {
    const query = db.query(`
      INSERT INTO jobs (id, name, status, startedAt, finishedAt, duration, queuedDuration, pipelineId, pipelineStatus, pipelineComputeMinutes, pipelineDuration, pipelineComplete)
      VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12)
`);
    query.run(
      job.id,
      job.name,
      job.status,
      job.startedAt,
      job.finishedAt,
      job.duration,
      job.queuedDuration,
      job.pipeline.id,
      job.pipeline.status,
      job.pipeline.computeMinutes,
      job.pipeline.duration,
      job.pipeline.complete ? 1 : 0,
    )
  });
}

async function runIngestionProcess() {
  let endCursor: string | undefined = undefined;
  let jobsCounter = 0;
  do {
    const { jobs, nextCursor } = await fetchJobs(endCursor);
    insertJobs(jobs);
    jobsCounter += PAGE_SIZE;
    console.log(`Retrieved ${jobsCounter} jobs`)
    endCursor = nextCursor;
  } while (endCursor && jobsCounter < JOB_LIMIT);
  db.close();
}

runIngestionProcess()
```

### 4) Ingesting the data

> Putting part 2 and 3 together, download and transformation only needs a `bun run`. The complete

Reviewer: ```suggestion
Merging steps 2 and 3, we executed our script with `bun run`, seamlessly ingesting the data. The complete
```

In getting the data, we can retrieve as many pages as we like. The maximum page size is 100, and we will
save the raw data in a file as a savepoint.
```typescript
const REPO_PATH = "path/to/repo"
const GITLAB_TOKEN = process.env["GITLAB_TOKEN"]
const PAGE_SIZE = 100

let jobs: object[] = []

async function fetchAllPages(endCursor?: string): Promise<void> {
  const apiUrl = 'https://gitlab.com/api/graphql';

  const query = `your_query`;

  const variables = endCursor ? { after: endCursor } : {};

  console.info(`Retrieving ${PAGE_SIZE} jobs after ${endCursor}`)

  const response = await fetch(apiUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${GITLAB_TOKEN}`
    },
    body: JSON.stringify({
      query,
      variables,
    }),
  });

  const result = await response.json() as any;

  const pageInfo = result.data.project.jobs.pageInfo;
  const nodes = result.data.project.jobs.nodes;

  jobs = jobs.concat(nodes);

  // Recursively fetch the next page if available
  if (pageInfo.endCursor && jobs.length < 20000) {
    console.log(`Got ${jobs.length} jobs, fetching more`)
    await fetchAllPages(pageInfo.endCursor);

>   }

Reviewer: Instead of saving raw data as an intermediate step, consider directly inserting the fetched data into an SQLite database. 

```ts

// Instead of accumulating jobs, insert them directly into the database
  db.serialize(() => {
    jobs.forEach((job) => {
      db.run(`INSERT INTO jobs VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`, [
        job.id, job.name, job.status, job.startedAt, job.finishedAt,
        job.duration, job.queuedDuration, job.pipeline.id,
        job.pipeline.status, job.pipeline.computeMinutes,
        job.pipeline.duration, job.pipeline.complete ? 1 : 0,
      ], (err) => {
        if (err) console.error('Insert error', err);
      });
    });
  });
```

Reviewer: ```ts
// Import necessary modules from Bun
import { SQLite } from 'bun:sqlite';
import { file } from 'bun:fs';

// Environment variables for configuration
const REPO_PATH = 'path/to/repo';
const GITLAB_TOKEN = Bun.env.GITLAB_TOKEN; // Use Bun.env for environment variables
const PAGE_SIZE = 100;
const DB_PATH = 'jobs_database.sqlite';
const GRAPHQL_ENDPOINT = 'https://gitlab.com/api/graphql';

// Interface for the structure of job data
interface JobData {
  id: string;
  name: string;
  status: string;
  startedAt: string;
  finishedAt: string;
  duration: number;
  queuedDuration: number;
  pipeline: {
    id: string;
    status: string;
    computeMinutes: number;
    duration: number;
    complete: boolean;
  };
}

// SQLite Database initialization
const db = new SQLite(DB_PATH);
db.query(`
  CREATE TABLE IF NOT EXISTS jobs (
    id TEXT PRIMARY KEY,
    name TEXT,
    status TEXT,
    startedAt TEXT,
    finishedAt TEXT,
    duration INTEGER,
    queuedDuration REAL,
    pipelineId TEXT,
    pipelineStatus TEXT,
    pipelineComputeMinutes REAL,
    pipelineDuration INTEGER,
    pipelineComplete INTEGER
  );
`);

// Function to fetch jobs data using GraphQL
async function fetchJobs(after?: string): Promise<{ jobs: JobData[]; nextCursor?: string }> {
  const query = `your GraphQL query here, including variables for pagination`;
  const variables = after ? { after, first: PAGE_SIZE } : { first: PAGE_SIZE };

  const response = await fetch(GRAPHQL_ENDPOINT, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${GITLAB_TOKEN}`,
    },
    body: JSON.stringify({
      query,
      variables,
    }),
  });

  const { data } = await response.json();
  const jobs: JobData[] = data.project.jobs.nodes;
  const nextCursor = data.project.jobs.pageInfo.endCursor;

  return { jobs, nextCursor };
}

// Function to insert fetched jobs into the SQLite database
function insertJobs(jobs: JobData[]) {
  jobs.forEach(job => {
    db.query(`
      INSERT INTO jobs (id, name, status, startedAt, finishedAt, duration, queuedDuration, pipelineId, pipelineStatus, pipelineComputeMinutes, pipelineDuration, pipelineComplete)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `, [
      job.id,
      job.name,
      job.status,
      job.startedAt,
      job.finishedAt,
      job.duration,
      job.queuedDuration,
      job.pipeline.id,
      job.pipeline.status,
      job.pipeline.computeMinutes,
      job.pipeline.duration,
      job.pipeline.complete ? 1 : 0,
    ]);
  });
}

// Main function to manage the ingestion process
async function runIngestionProcess() {
  let endCursor: string | undefined = undefined;
  do {
    const { jobs, nextCursor } = await fetchJobs(endCursor);
    insertJobs(jobs);
    endCursor = nextCursor;
  } while (endCursor);
  db.close();
}

// Run the ingestion process
runIngestionProcess().catch(console.error);

```

Reviewer: Cool idea could be also to utilise the new bun shell. Then we could have a super small cli tool for executing it, which we could release as a separate cli tool.

https://bun.sh/docs/runtime/shell

Author: Updated the code, thanks 👍🏻 

About the cli interface, I don't care about it as this is just an ad-hoc script, but feel free to add it.

  CREATE TABLE IF NOT EXISTS jobs (
    id TEXT PRIMARY KEY,
    name TEXT,
    status TEXT,
    startedAt TEXT,
    finishedAt TEXT,
    duration INTEGER,
    queuedDuration REAL,
    pipelineId TEXT,
    pipelineStatus TEXT,
    pipelineComputeMinutes REAL,
    pipelineDuration INTEGER,
    pipelineComplete INTEGER
  );
`.replaceAll("\n", "");

async function createQueries() {
  const jsonJobs = Bun.file("raw_data.json")
  const jobs: JobData[] = JSON.parse(await jsonJobs.text())

  let output = [createTable]
  output = output.concat(jobs.map((job) => {
    return denormalizeToSQLiteStatements(job)
  }))

  await Bun.write("queries.sql", output.join("\n"))
}
```

### 4) Ingesting the data

After running step 2 and 3, ingesting the transformed data is easy as pie:
```sh
sqlite3 test.sqlite < queries.sql
```

For our 20000 jobs, which is the amount of jobs triggered by a 4 man team over 5 months in one
repository, the resulting sqlite file is 5MB. The bottleneck here is more with the fetching of the
job data, but in our two-step process we don't have to do this too many times.

### 5) Preparing Grafana

Next up is starting a local grafana instance.

```sh
mkdir data

> docker run -d -p 3000:3000 --name=grafana \

Reviewer: Maybe we could use a provided docker-compose file. Could reduce the amount of shell cmds.

```yml
version: '3.8'
services:
  grafana:
    image: grafana/grafana-enterprise
    environment:
      - GF_INSTALL_PLUGINS=frser-sqlite-datasource
    volumes:
      - ./data:/var/lib/grafana
      - ./gitlab_jobs.sqlite:/var/lib/grafana/gitlab_jobs.sqlite
    ports:
      - "3000:3000"
```

```bash
docker-compose up -d
```

Author: I guess I favour having simply the docker run command, which I copied from the grafana docs. Don't care too much about this though, feel free to change it.
